2024/BIN-2024-0003.md

```diff
@@ -0,0 +1,124 @@
+    * If the signature is the empty vector:
+        * For `OP_CHECKSIGFROMSTACKVERIFY`, the script MUST fail and terminate immediately.
+        * For `OP_CHECKSIGFROMSTACK`, an empty vector is pushed onto the stack, and execution continues with the next opcode.
+    * If the signature is not the empty vector:
+        * For tapscript 0xc0, the opcode is counted towards the sigops budget as described in BIP341.
+        * For legacy and segwit v0, the opcode is counted towards the sigops limit, as described in BIP141
+        * For `OP_CHECKSIGFROMSTACKVERIFY`, execution continues without any further changes to the stack.
+        * For `OP_CHECKSIGFROMSTACK`, a 1-byte value 0x01 is pushed onto the stack.
+
+## Design Considerations
+
+1. Message hashing: ECDSA requires the message to be securely hashed to 32-bytes before being used in the signing protocol, so a SHA256 hash is taken before the message is passed for ECDSA signing. BIP340 is compatible with any size of message and does not require it to be a securely hashed input, so the message is not hashed prior to BIP340 signing.
+2. Verify NOP upgrade: To bring both ECDSA and BIP340 signing to bitcoin script, a NOP upgrade path was chosen for `OP_CHECKSIGFROMSTACKVERIFY`. This necessarily means leaving the 3 arguments on the stack when executing `OP_CHECKSIGFROMSTACKVERIFY`. Scripts will need to drop or otherwise manage these stack elements.
+3. Add/multisig: No concession is made to `OP_CHECKMULTISIG` or `OP_CHECKSIGADD` semantics with `OP_CHECKSIGFROMSTACK(VERIFY)`. In Tapscript, add semantics can be implemented with 1 additional vByte per key (`OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD OP_TOALTSTACK`).
+4. Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for future bitcoin upgrades, such as `OP_CAT`.
+5. BIP118-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing OP_INTERNALKEY or separately introducing that key type for all Tapscript signature checking operations in a separate change.
+6. Unknown key lengths: The semantics of other signature checking opcodes in their respective script types (legacy, segwit-v0, tapscript-c0) are applied.
+
+## Resource Limits
+
+These opcodes are treated identically to other signature checking opcodes and
+count against the various sigops limits and budgets in their respective script
+types.
+
+## Motivation
+
+### LN Symmetry
+
+When combined with `OP_CHECKTEMPLATEVERIFY` (BIP119/CTV),
+`OP_CHECKSIGFROMSTACK` (CSFS) can be used in Lightning Symmetry channels.
+The construction `OP_CHECKTEMPLATEVERIFY <pubkey> OP_CHECKSIGFROMSTACK` is
+logically equivalent to `<bip118_pubkey> OP_CHECKSIG` and a signature over
+`SIGHASH_ALL|SIGHASH_ANYPREVOUTANYSCRIPT`. The `OP_CHECKSIGFROMSTACK`
+construction is 8 vBytes larger.
+
+### Delegation
+
+Using a script like:
+`OP_DUP <pubkey> OP_CHECKSIGFROMSTACK OP_DROP OP_CHECKSIG`
```

Comment:
I think this would be better as `OVER CHECKSIGVERIFY <p> CHECKSIGFROMSTACK` with the witness data being `<sig-of-q-by-p> <q> <sig-by-q>`. I don't think it works as-is.

Reply:
I agree it doesn't work as written. I have a more useful version floating around that I'll replace it with, thanks!